static_frame/core/index_hierarchy.py

```diff
@@ -456,211 +719,273 @@ def _from_type_blocks(cls: tp.Type[IH],
-
-        if isinstance(levels, IndexHierarchy):
-            if not blocks is None:
-                raise ErrorInitIndex('cannot provide blocks when initializing with IndexHierarchy')
-            index_level = levels._levels
-            # if cache is updated, can get blocks
-            if not levels._recache:
-                self._blocks = levels._blocks.copy()
-            # transfer name if not given as arg
-            if name is NAME_DEFAULT:
-                name = levels.name
+        if isinstance(indices, IndexHierarchy):
+            if indexers is not EMPTY_ARRAY_INT:
+                raise ErrorInitIndex(
+                    'indexers must not be provided when copying an IndexHierarchy'
+                )
+            if blocks is not None:
+                raise ErrorInitIndex(
+                    'blocks must not be provided when copying an IndexHierarchy'
+                )
 
-        elif isinstance(levels, IndexLevel):
-            index_level = levels
-            if not blocks is None:
-                self._blocks = blocks if own_blocks else blocks.copy()
+            if indices._recache:
+                indices._update_array_cache() # pragma: no cover
+
+            self._indices = [
+                mutable_immutable_index_filter(self.STATIC, index)
+                for index in indices._indices
+            ]
+            self._indexers = indices._indexers
+            self._name = name if name is not NAME_DEFAULT else indices._name
+            self._blocks = indices._blocks
+            self._values = indices._values
+            self._map = indices._map
+            return
+
+        if not (indexers.__class__ is np.ndarray and not indexers.flags.writeable):
+            raise ErrorInitIndex('indexers must be a read-only numpy array.')
+
+        if not all(isinstance(index, Index) for index in indices):
+            raise ErrorInitIndex("indices must all Index's!")
+
+        if len(indices) <= 1:
+            raise ErrorInitIndex('Index Hierarchies must have at least two levels!')
+
+        self._indices = [
+            mutable_immutable_index_filter(self.STATIC, index) # type: ignore
+            for index in indices
+        ]
+        self._indexers = indexers
+        self._name = None if name is NAME_DEFAULT else name_filter(name)
 
+        if blocks is not None:
+            if own_blocks:
+                self._blocks = blocks
+            else:
+                self._blocks = blocks.copy()
         else:
-            raise NotImplementedError(f'no handling for creation from {levels}')
+            self._blocks = self._create_blocks_from_self()
 
-        if self.STATIC and index_level.STATIC:
-            self._levels = index_level
-        else: # must deepcopy IndexLevels if not IndexHierarchy not static
-            self._levels = index_level.to_index_level(
-                    cls=self._LEVEL_CONSTRUCTOR
-                    )
+        self._values = self._blocks.values
```

Comment:
This will force type consolidation: is that acceptable?

Reply:
Yeah - this is only used by the `values` property.

Reply:
So we are eagerly-caching `values`; but is that necessary? Is it better to be lazy and/or not hold on to `values` and simply create `values` on the fly when requested (exactly what `TypeBlocks` does).

Reply:
Initially I saw it as an opporunity for improvement, but I think that was before I fully realized that `values` isn't really a core feature of the class. I'll revert it back to be created when requested.